protocol/contracts/governance/ElderElection.sol

```diff
@@ -0,0 +1,177 @@
+
+    /// @notice endorsements by a given voter
+    mapping(address => uint256[]) public endorsementsBy;
+
+    /// @notice the NFT contract for templars
+    ERC721 public templars;
+
+    /// @notice Token which nomination fee is charged in
+    IERC20 public immutable nominationFeeToken;
+
+    /// @notice Nomination fee amount (in feeToken)
+    uint256 public nominationFee;
+
+    /// @notice root hash of off-chain merkle tree recording
+    /// temple roles ie :(discordId, maxTempleRole)
+    bytes32 public templarRolesMerkleRoot;
+
+    /// @notice Nonces used in relayed voting requests
+    mapping(address => Counters.Counter) public _nonces;
+
+
+    constructor(
+        ERC721 _templars,
+        IERC20 _nominationFeeToken,
+        uint256 _nominationFee
+    ) EIP712("ElderElection", "1") {
+        templars = _templars;
+        nominationFeeToken = _nominationFeeToken;
+        nominationFee = _nominationFee;
+    }
+
+    /**
+     * @notice Change the nomination fee
+     * 
+     * Designed to be low, but a defense in depth to stop peeps putting up a nomination
+     * for the lolz.
+     */
+    function setNominationFee(uint256 _nominationFee) external onlyOwner {
+        nominationFee = _nominationFee;
+    }
+
+    /**
+     * @notice Update the merkle tree root for temple roles.
+     */
+    function setTemplarRolesMerklRoot(bytes32 _templarRolesMerkleRoot) external onlyOwner {
+        templarRolesMerkleRoot = _templarRolesMerkleRoot;
+    }
+
+    /**
+     * @notice Nominate the given discord ID a candidate.
+     * @dev The given address must hold the NFT for the specified discord id
+     */
+    function nominate(
+        address account,
+        uint256 discordId
+    ) external {
+        if (templars.ownerOf(discordId) != account) {
+            revert NotFromTemplar(account, discordId);
+        }
+
+        if (!candidates[discordId]) {
+
+          if (nominationFee > 0) {
+              nominationFeeToken.safeTransferFrom(account, address(this), nominationFee);
+          }
+          candidates[discordId] = true;
+          emit UpdateNomination(discordId, candidates[discordId]);
+        }
+    }
+
+    /**
+     * @notice Remove the given discord ID as a candidate
+     * @dev The given address must hold the NFT for the specified discord id
+     */
+    function resign(
+        address account,
+        uint256 discordId
+    ) external {
+        if (templars.ownerOf(discordId) != account) {
+            revert NotFromTemplar(account, discordId);
+        }
+
+        if (candidates[discordId]) {
+          candidates[discordId] = false;
+          emit UpdateNomination(discordId, candidates[discordId]);
+        }
+    }
+
+    function setEndorsements(uint256[] memory discordIds) external {
+        _setEndorsements(msg.sender, discordIds);
+    }
+
+    function _setEndorsements(address account, uint256[] memory discordIds) internal {
+        endorsementsBy[account] = discordIds;
```

Comment:
docs recommend to use `uint256[] calldata discordIds`
https://docs.soliditylang.org/en/v0.8.4/types.html?highlight=calldata#data-location

I wonder if we should put a bound on the number of allowed endorsements (eg 10?), just so this storage isn't ridiculous?

Probably need a custom getter to get all discordIds back in one hit. With the default getter, caller would need to provide an array index to get 1-by-1 (and so would need to know then length of the array)
```
function getEndorsementsBy(address account) external returns (uint256[] calldata) {
    return endorsementsBy[account];
}
```

Reply:
Yes - will limit endorsements to active candidates.